Integrations/Digital_Defense_FrontlineVM/Digital_Defense_FrontlineVM.py

```diff
@@ -0,0 +1,795 @@
+import signal
+import math
+import re
+from datetime import datetime, tzinfo, timedelta
+from typing import List, Dict, Any    # throwing error: "ImportError: No module named typing"
+
+# Params:
+API_TOKEN = demisto.params().get('apiToken')
+INCIDENT_VULN_MIN_SEVERITY = demisto.params().get('incidentSeverity')
+INCIDENT_FREQUENCY = demisto.params().get('incidentFrequency')
+
+
+def get_base_url():
+    ''' Removes forward slash from end of input '''
+    url = demisto.params().get('frontlineURL')
+    url = re.sub(r'\/$', '', url)
+    return url
+
+
+# Endpoints:
+BASE_URL = get_base_url()
+VULN_ENDPOINT = BASE_URL + "/api/scanresults/active/vulnerabilities/"
+HOST_ENDPOINT = BASE_URL + "/api/scanresults/active/hosts/"
+SCAN_ENDPOINT = BASE_URL + "/api/scans/"
+
+# FrontlineVM (FVM) request header for API calls:
+FVM_HEADER = {'Authorization': 'Token ' + str(API_TOKEN)}
+
+# Minimum time to timeout functions (5 mins)
+MIN_TIMEOUT = 300   # seconds
+
+# HEADERS:
+VULN_DATA_HEADERS = ['vuln-id',
+                     'hostname',
+                     'ip-address',
+                     'vuln-title',
+                     'date-created',
+                     'ddi-severity',
+                     'vuln-info']
+
+HOST_HEADERS = ['ID',
+                'Hostname',
+                'IP',
+                'DNSHostname',
+                'MAC',
+                'OS',
+                'OSType',
+                'CriticalVulnCount']
+
+
+'''HELPER FUNCTIONS'''
+
+
+class EndOfTime(Exception):
+    ''' Raised when functions timeout '''
+    pass
+
+
+def function_timeout(signum, frame):
+    '''
+    Used to raise EndOfTime exception for timeout functions.
+    '''
+    raise EndOfTime('Function has timed out')
+
+
+def get_function_timeout_time(data_count):
+    '''
+    Returns time (in seconds) to timeout function
+    based upon the amount of data to pull.
+    '''
+    timeout_time = math.ceil(data_count / 2)
+    if timeout_time < MIN_TIMEOUT:
+        timeout_time = MIN_TIMEOUT
+    return timeout_time
+
+
+def get_all_data(first_page):
+    '''
+    Retrieves all data if multiple pages are present in API request.
+    '''
+    request_url = first_page.get('next')
+    have_all_data = False
+    current_data = {}   # type: Dict[str, Any]
+    all_data = []       # type: List[Dict]
+    while not have_all_data:
+        resp = requests.get(url=request_url, headers=FVM_HEADER, timeout=30)
+        if not resp.ok:
+            msg = "FrontlineVM get_all_data -- status code: " + str(resp.status_code)
+            demisto.debug(msg)
+        resp.raise_for_status()
+        current_data = json.loads(resp.text)
+        all_data.extend(current_data.get('results', []))
+        if current_data.get('next'):
+            request_url = current_data.get('next')
+        else:
+            have_all_data = True
+    return all_data
+
+
+def get_fvm_data(request_url, **kwargs):
+    ''' Retrieves data from FrontlineVM API '''
+    data = []   # type: List
```

Comment:
Which data does this function get? let's be more specific

Reply:
This date isn't specific because depending on the request_url this method can pull either asset data, vuln data, or scan data. This is used as a generic function to pull Frontline Vulnerability Management data from our API.